### PR TITLE
Runtime: fix hex float parsing with underscore in exponent

### PR DESCRIPTION
**Pull request by hhugo**
Following the format described in https://caml.inria.fr/pub/docs/manual-ocaml/lex.html.

Note that the ocaml lexer is more restrictive than float_of_string. In particular, float_of_string accepts leading underscore in the exponent which is rejected in the ocaml lexer.

TODO:
- [x] add some tests with null bytes in the input
- [x] add some tests showing that underscores are accepted everywhere.

### CODE REVIEW

**Comment by hhugo** [comment]
Here is what would happen before this PR.
```
$ echo "0x1p1_1" > a.ml && ocamlc a.ml
Fatal error: exception Failure("float_of_string")
Raised by primitive operation at file "bytecomp/symtable.ml", line 209, characters 42-61
Called from file "bytecomp/symtable.ml", line 233, characters 39-55
Called from file "list.ml", line 106, characters 12-15
Called from file "bytecomp/symtable.ml", line 232, characters 2-87
Called from file "bytecomp/bytelink.ml", line 367, characters 14-47
Re-raised at file "bytecomp/bytelink.ml", line 386, characters 4-11
Called from file "driver/main.ml", line 199, characters 6-67
Re-raised at file "parsing/location.ml", line 479, characters 14-25
Re-raised at file "parsing/location.ml", line 479, characters 14-25
Re-raised at file "parsing/location.ml", line 479, characters 14-25
Re-raised at file "parsing/location.ml", line 479, characters 14-25
Re-raised at file "parsing/location.ml", line 479, characters 14-25
Re-raised at file "parsing/location.ml", line 479, characters 14-25
Called from file "parsing/location.ml" (inlined), line 484, characters 31-61
Called from file "driver/main.ml", line 203, characters 4-35
Called from file "driver/main.ml", line 207, characters 2-9
EXIT STATUS 2
```

**Comment by xavierleroy** [comment]
I was wondering whether we really want underscores in exponents of FP numbers, but it's probably too late to change our minds about this, as they've been supported for ages in the case of decimal FP notation, e.g.
```
# 1e3_00;;
- : float = 1e+300
#  float_of_string "1e3_00";;
- : float = 1e+300
```
I still find the "copy string while removing underscores" code inelegant, but it's better than crashing the compiler.

**Comment by gasche** [comment]
Could we have a fast path that does not do anything if no `_` are present? If we start by counting the number of `_`, this may simplify the size-calculation logic in the slow path as well.

**Comment by hhugo** [comment]
> I still find the "copy string while removing underscores" code inelegant, but it's better than crashing the compiler.

This implementation was copied and adapted from `caml_float_of_string`.

Regarding the fast path, based on the following point, I'm tempted to do nothing.
- The non-hexadecimal path already does the "copy string while removing underscores"
- Hexadecimal floats are pretty uncommon

**Comment by xavierleroy** [comment]
I'm not too concerned about speed, more concerned about code duplication.  Perhaps float_of_string should always remove underscores first, then handle decimal and hex floats separately.  I'll try to read the whole code again to form an opinion.

**Comment by gasche** [comment]
It may have been a mistake to talk of fast/slow path, but I think having a direct path in the simple case is also a win for correctness. You have modified more than 10 lines of C code doing index computations, read and write in a fixed-size array, so that's above the limit where I feel reasonably confident that nothing goes wrong by just eyeballing the code. On the other hand, I could check a "simple path" implementation to easily become very confident that all existing use-cases (without underscores) are completely untouched by the change -- and thus be more at ease when reading the "complex path" code.

**Comment by hhugo** [comment]
I gave a try at @gasche's suggestion. See last commits

File: runtime/floats.c [hunk 1]
```diff
@@ -372,51 +370,63 @@ static int caml_float_of_hex(const char * s, double * res)
...
-    char c = *src++;
-    if (c != '_') *dst++ = c;
-  }
-  *dst = 0;
-  if (dst == buf) goto error;
+  else {
+    char * end;
 #if defined(HAS_STRTOD_L) && defined(HAS_LOCALE)
-  d = strtod_l((const char *) buf, &end, caml_locale);
+    *res = strtod_l(src, &end, caml_locale);
```
Comment (hhugo) [inline runtime/floats.c]: This diff is not correct, `src` was potentially moved already and no longer contains the sign.

**Comment by hhugo** [comment]
`float_of_string "0x1\000a"` currently returns some result. With this PR, one would raise an exception, similar to what happen with the decimal notation.

File: runtime/floats.c (hunk 1 above)
Comment (hhugo) [reply runtime/floats.c]: fixed

**Comment by xavierleroy** [comment]
Your code was getting too complicated for me to review, so I took the liberty of pushing straight onto your branch the simpler fix I had in mind, namely remove underscores first, then parse as hex or as decimal.  Again, I'm not worried about the possible run-time costs, but I am concerned about code complexity.

I'll give this PR a round of CI precheck next.

**Review by gasche — APPROVED** [review]
I reviewed @xavierleroy's new implementation and I believe that it is correct -- sorry for not noticing the simplification myself.

Optionally, the whole code could be further simplified/streamlined by:
- merging the return continuations of the of hexademical and decimal cases (currently the `caml_stat_free` and `caml_copy_double` logics are duplicated)
- moving the decimal logic (do we have strod_l or not?) to its own `caml_float_of_dec` function, so that the two branches are at the same abstraction level within the main function.

**Comment by hhugo** [comment]
> the simpler fix I had in mind, namely remove underscores first, then parse as hex or as decimal.

My last was doing that already, no?

> merging the return continuations of the of hexademical and decimal cases

This simplication was part of my last version

**Comment by xavierleroy** [comment]
@hhugo: reviewing code is harder than writing it, and your latest diff was so large that I certainly won't review it.  From this point on, I'll write the code.  Please don't feel offended, I'm just trying to use my time as efficiently as I can.

**Comment by xavierleroy** [comment]
This said, both @hhugo's and my latest code accept things like `_0x1p2` or `0_x1p2`, which can be questioned.

**Comment by xavierleroy** [comment]
I give up on the idea of rejecting e.g. `0_x1p2`.  Let's be tolerant in what we accept...

I just pushed a cleanup of the code, with more sharing of continuations as suggested by @gasche.

**Comment by hhugo** [comment]
@xavierleroy, should we be tolerant in what we accept for ints as well?

**Comment by xavierleroy** [comment]
> should we be tolerant in what we accept for ints as well?

For me a reasonable goal for `int_of_string` and float_of_string is that they match OCaml's lexical conventions for integer literals and FP literals.  For `string_of_float` we accept more than just OCaml FP literals, because we don't control what `strtod` does.  Checking the string before calling `strtod` would be a lot of code for a small benefit.

But for `int_of_float`  I think the current implementation accepts all OCaml integer literals and only OCaml integer literals.  I'm not sure we want to accept more.  Did you have something specific in mind?

**Review by gasche — APPROVED** [review]
I have looked at the current patch and I believe that the code is correct.

**Comment by hhugo** [comment]
@xavierleroy , nothing specific in mind. I'm fine with the current diff.

**Comment by xavierleroy** [comment]
OK, this is merged in trunk.  Should it be cherry-picked to 4.08?  @gasche and I have cold feet (because it is very late in 4.08's release cycle) but @damiendoligez can decide.

**Comment by damiendoligez** [comment]
@hhugo how did you find this bug? I have a hard time believing that it's big enough to justify cherry-picking.

**Comment by hhugo** [comment]
I discovered the bug more or less randomly. I would not push hard to have a fix in 4.08.
